### ISSUE TITLE
Kubernetes Helm Testing

### ISSUE THREAD

**Issue by odatskov** [singularityhub/sregistry#317]
Dear Developers,

I am currently testing running a Singularity registry on Kubernetes resources. I have ported the docker compose you have into a [basic Helm chart](https://github.com/odatskov/helm-sregistry). Three questions I have:

- The obvious one: is there already a chart or are you working on one (so as not to duplicate effort)?
-  Based on [storage docs](https://singularityhub.github.io/sregistry/docs/install/server#storage): for recent container versions is /var/www/images still needed to be a persistent storage place?  
- worker redis dependency: hardcoded to be a local service or can also be configured with external?

**Comment by vsoch** [comment]
It’s so great that you are doing this! There isn’t a chart and it would be great to have one. The images folder is kept for backwards compatibility, and I believe might still be used if you upload a container directly in the interface (do you have it running and could test?) And I also don’t see why the redis worked could be configured to be external.

Really looking forward to checking this out!

**Comment by odatskov** [comment]
Most excellent! Thank you for such a quick response. Indeed I am running on local K8s infra and will definitely check the interactive upload.

**Comment by odatskov** [comment]
Tested with persistent storage and GitHub auth. Singularity shub pull seems fine, however the push is giving a bit of trouble: singularity doesn't like the tokenfile (signed token?) and with sregistry push I am getting an exception of "certificate verify failed" (even with SREGISTRY_REGISTRY_NOHTTPS=true). 

In regards to previous comments:
- Interactive webui push: indeed the sif image can be found in /var/www/images/<collection>
- Redis custom url I found can be set through REDIS_URL environment.
- settings/config.py [duplicate](https://github.com/singularityhub/sregistry/blob/master/shub/settings/config.py#L72) of DISABLE_BUILDING or expected?

**Comment by vsoch** [comment]
Ah, I bet I know the issue! For the library endpoint you actually can't use http - so the way that I test locally is by compiling singularity with remote.go set to http. Just change [this line](https://github.com/hpcng/singularity/blob/master/internal/pkg/remote/remote.go#L237) from https to http, and then build. The instructions are in the docs [here](https://singularityhub.github.io/sregistry/docs/client#singularity-push). The disable_building repeat is an error! I have a PR open now and I'll remove the extra one.

**Comment by vsoch** [comment]
Any updates on working on a chart?

**Comment by odatskov** [comment]
I am indeed testing the chart on our infra. Will have results in a bit.

**Comment by ifelsefi** [comment]
Any update with the chart?  I would like to use this as well.

**Comment by vsoch** [comment]
I suspect @odatskov is not working on it anymore. @odatskov - can you share what you started so possibly someone else can take over? And if you had any kind of issues, tell us and maybe we can help?

**Comment by ifelsefi** [comment]
I forked it.

`quay.io/vanessa/sregistry_nginx:1.1.34` `fails due to 2021/04/06 19:32:28 [emerg] 1#1: invalid number of arguments in "client_body_timeout" directive in /etc/nginx/conf.d/default.conf:7`

Deleted that entirely then new errors in same container.   Would be nice if @odatskov would share feedback but seems to be AFK for a long time.

**Comment by vsoch** [comment]
You don't technically need sregistry_nginx anymore - that was when we had a custom nginx module to do uploads. The current sregistry using the sylabs library API with a minio storage backend. So I'd remove that image and use some standard nginx.

**Comment by vsoch** [comment]
There is also some "wisdom" on stack overflow if you want to try debugging it (without removing it) https://serverfault.com/questions/786982/nginx-conditional-proxy-invalid-number-of-arguments-in-try-files-directive

**Comment by ifelsefi** [comment]
Hmm ok I am using a nginx image with your release `1.1.34`.

The uwsgi container gives me this django error:

```Forbidden (Permission denied): /
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.5/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.5/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.5/site-packages/ratelimit/decorators.py", line 29, in _wrapped
    raise Ratelimited()
ratelimit.exceptions.Ratelimited
```

I will keep checking the config as this would save us a lot of time.  It is a pain to download images from our private docker registry.

**Comment by ifelsefi** [comment]
So `uwsgi` dies due to permission denied / error but the k8s health check does not know about that.   So the pod stays up, in bad state, and `nginx` keeps trying to connect.   My nginx health checks do detect a bad state when it won't proxy to upstream and so that container keeps restarting.   So need to fix the permission denied issue with "/".

**Comment by vsoch** [comment]
That is rate limiting enforced by the server. You likely would want to change or remove it for this use case. We don't have any official K8 deployment, but it would be good if you get something working to write it up, include configs here, and then either note or have an automated way to disable or change the rate limiting.

**Comment by ifelsefi** [comment]
Thank you.   I wonder if that would cause the permission denied "/" error?   I will push my changes to the fork.

**Comment by vsoch** [comment]
Yes when the rate limit is exceeded it raises a permission denied, 403.

```
[pid: 63|app: 0|req: 51/55] 10.69.24.84 () {32 vars in 361 bytes} [Thu Jun 17 09:07:33 2021] GET / => generated 22 bytes in 3 msecs (HTTP/1.1 403) 3 headers in 100 bytes (1 switches on core 3)
Forbidden (Permission denied): /
```

**Comment by ifelsefi** [comment]
Ah, [this is an option already](https://github.com/ifelsefi/helm-sregistry/blob/master/sregistry/templates/sregistry-cm.yaml#L152) with the helm chart created by the author.

```    # Rate Limits

    VIEW_RATE_LIMIT = {{ .Values.config.viewRateLimit | quote }}  # The rate limit for each view, django-ratelimit, "50 per day per ipaddress)
    VIEW_RATE_LIMIT_BLOCK = (
        {{ .Values.config.viewRateBlock }}  # Given that someone goes over, are they blocked for the period?
    )
```

Thanks for pointing me in the right direction!

**Comment by vsoch** [comment]
Definitely!
